cli/pcluster/config/param_types.py

```diff
@@ -0,0 +1,1202 @@
+
+
+class DesiredSizeParam(IntParam):
+    """
+    Class to manage both the initial_queue_size and desired_vcpus configuration parameters.
+
+    We need this class since the same CFN input parameter "DesiredSize" is populated
+    from the "desired_vcpus" parameter when the scheduler is awsbatch and
+    from "initial_queue_size" when the scheduler is a traditional one.
+    """
+
+    def _init_from_cfn(self, cfn_params=None, cfn_value=None):
+        """Initialize param value by parsing the right CFN input according to the scheduler."""
+        cfn_converter = self.map.get("cfn", None)
+        if cfn_converter and cfn_params:
+            cfn_value = get_cfn_param(cfn_params, cfn_converter) if cfn_converter else "NONE"
+            # initialize the value from cfn or from map according to the scheduler
+            if get_cfn_param(cfn_params, "Scheduler") == "awsbatch":
+                if self.key == "initial_queue_size":
+                    self._init_from_map()
+                elif self.key == "desired_vcpus":
+                    self.value = self.get_value_from_string(cfn_value)
+            else:
+                if self.key == "initial_queue_size":
+                    self.value = self.get_value_from_string(cfn_value)
+                elif self.key == "desired_vcpus":
+                    self._init_from_map()
+        elif cfn_value:
+            self.value = self.get_value_from_string(cfn_value)
+        else:
+            self._init_from_map()
+
+    def to_cfn(self):
+        """Convert parameter to CFN representation."""
+        cfn_params = {}
+
+        cluster_config = self.pcluster_config.get_section(self.section_key)
+        if cluster_config.get_param_value("scheduler") == "awsbatch":
+            cfn_value = cluster_config.get_param_value("desired_vcpus")
+            cfn_params.update({self.map.get("cfn"): str(cfn_value)})
+        else:
+            cfn_value = cluster_config.get_param_value("initial_queue_size")
+            cfn_params.update({self.map.get("cfn"): str(cfn_value)})
+
+        return cfn_params
+
+
+class MaxSizeParam(IntParam):
+    """
+    Class to manage both the max_queue_size and max_vcpus configuration parameters.
+
+    We need this class since the same CFN input parameter "MaxSize" is populated
+    from the "max_vcpus" parameter when the scheduler is awsbatch and
+    from "max_queue_size" when the scheduler is a traditional one.
+    """
+
+    def _init_from_cfn(self, cfn_params=None, cfn_value=None):
+        """Initialize param value by parsing the right CFN input according to the scheduler."""
+        cfn_converter = self.map.get("cfn", None)
+        if cfn_converter and cfn_params:
+            cfn_value = get_cfn_param(cfn_params, cfn_converter) if cfn_converter else "NONE"
+            # initialize the value from cfn or from map according to the scheduler
+            if get_cfn_param(cfn_params, "Scheduler") == "awsbatch":
+                if self.key == "max_queue_size":
```

Comment:
to simplify all these conditional statements can we just do an _init_from_map() at the very beginning (or in the base class) and then override the value when necessary? (same in all other functions that have a similar logic)

Reply:
Done